File: min0_core_forth_root.rb
# frozen_string_literal: true

require "digest"
require "json"
require_relative "min0_core_forth_install"

module Min0CoreForth
  ROOT_POLICY_FORMAT = "min0-core-forth-root-policy"
  ROOT_POLICY_VERSION = 1
  ROOT_POLICY_DOMAIN = "MIN0-CORE-FORTH-ROOT-POLICY-R0\0".b
  ROOT_STATE_DOMAIN = "MIN0-CORE-FORTH-ROOT-STATE-R0\0".b
  ROOT_POLICY_INSTALL_STEPS = [
    "erase-inactive-root-state", "write-root-policy-chain", "seal-root-state"
  ].freeze
  ROOT_KEY_ID_PATTERN = /\A[a-z0-9][a-z0-9._-]{0,63}\z/

  class RootPolicyError < StandardError; end
  RootStateSlot = Struct.new(:chain, :checksum, keyword_init: true)

  module RootPolicy
    module_function

    def field(mapping, name)
      return nil unless mapping.respond_to?(:key?)
      return mapping[name] if mapping.key?(name)

      mapping[name.to_s]
    end

    def key_id(value)
      unless value.is_a?(String) && value.match?(ROOT_KEY_ID_PATTERN)
        raise RootPolicyError, "root key_id is malformed"
      end

      value
    end

    def public_key(value)
      unless value.is_a?(String) && value.match?(/\A[0-9a-f]{64}\z/)
        raise RootPolicyError, "root public key must be 32 bytes of lowercase hex"
      end

      [value].pack("H*")
    end

    def signature(value)
      unless value.is_a?(String) && value.match?(/\A[0-9a-f]{128}\z/)
        raise RootPolicyError, "root policy signature must be 64 bytes of lowercase hex"
      end

      [value].pack("H*")
    end

    def normalize_roots(roots)
      raise RootPolicyError, "root entries must be an Array" unless roots.is_a?(Array)

      seen = {}
      normalized = roots.map do |entry|
        raise RootPolicyError, "root entry must be a Hash" unless entry.is_a?(Hash)
        unless entry.keys.map(&:to_s).sort == %w[key_id public_key_hex status]
          raise RootPolicyError, "root entry fields are malformed"
        end
        normalized_id = key_id(field(entry, :key_id))
        raise RootPolicyError, "duplicate root key_id" if seen[normalized_id]

        seen[normalized_id] = true
        normalized_public = public_key(field(entry, :public_key_hex))
        status = field(entry, :status)
        unless %w[active retired].include?(status)
          raise RootPolicyError, "root status must be active or retired"
        end
        {
          key_id: normalized_id,
          public_key_hex: normalized_public.unpack1("H*"),
          status: status
        }
      end
      normalized.sort_by! { |entry| entry[:key_id] }
      unless normalized.any? { |entry| entry[:status] == "active" }
        raise RootPolicyError, "root policy must retain an active root"
      end
      normalized
    end

    def normalize_signatures(signatures)
      unless signatures.is_a?(Array)
        raise RootPolicyError, "root signatures must be an Array"
      end

      seen = {}
      signatures.map do |entry|
        raise RootPolicyError, "root signature entry must be a Hash" unless entry.is_a?(Hash)
        unless entry.keys.map(&:to_s).sort == %w[key_id signature_hex]
          raise RootPolicyError, "root signature entry fields are malformed"
        end
        normalized_id = key_id(field(entry, :key_id))
        raise RootPolicyError, "duplicate root signature key_id" if seen[normalized_id]

        seen[normalized_id] = true
        {
          key_id: normalized_id,
          signature_hex: signature(field(entry, :signature_hex)).unpack1("H*")
        }
      end.sort_by { |entry| entry[:key_id] }
    end

    def payload(policy)
      rows = normalize_roots(field(policy, :roots)).map do |entry|
        [entry[:key_id], entry[:public_key_hex], entry[:status]]
      end
      JSON.generate(
        [
          field(policy, :format), field(policy, :version), field(policy, :epoch),
          field(policy, :previous_policy_sha256), rows
        ]
      )
    end

    def digest(policy)
      Digest::SHA256.hexdigest(payload(policy))
    end

    def message(policy)
      ROOT_POLICY_DOMAIN + [digest(policy)].pack("H*")
    end

    def root_map(roots, status = nil)
      roots.filter_map do |entry|
        next unless status.nil? || entry[:status] == status

        [entry[:key_id], [entry[:public_key_hex]].pack("H*")]
      end.to_h
    end

    def build(epoch, roots, signer_private_keys, previous_policy: nil)
      normalized_epoch = Generation.validate(epoch, "root policy epoch")
      policy = {
        format: ROOT_POLICY_FORMAT,
        version: ROOT_POLICY_VERSION,
        epoch: normalized_epoch,
        previous_policy_sha256: previous_policy.nil? ? nil : digest(previous_policy),
        roots: normalize_roots(roots),
        signatures: []
      }
      policy[:signatures] = signer_private_keys.sort_by { |id, _key| id }.map do |id, key|
        normalized_id = key_id(id)
        begin
          signed = key.sign(nil, message(policy))
        rescue OpenSSL::PKey::PKeyError, TypeError, ArgumentError, NoMethodError => e
          raise RootPolicyError, "invalid root signer private key: #{e.message}"
        end
        { key_id: normalized_id, signature_hex: signed.unpack1("H*") }
      end
      policy
    end

    def validate_shape(policy)
      raise RootPolicyError, "root policy must be a Hash" unless policy.is_a?(Hash)
      unless field(policy, :format) == ROOT_POLICY_FORMAT
        raise RootPolicyError, "unsupported root policy format"
      end
      unless field(policy, :version) == ROOT_POLICY_VERSION
        raise RootPolicyError, "unsupported root policy version"
      end
      epoch = Generation.validate(field(policy, :epoch), "root policy epoch")
      previous = field(policy, :previous_policy_sha256)
      unless previous.nil? || (previous.is_a?(String) && previous.match?(/\A[0-9a-f]{64}\z/))
        raise RootPolicyError, "previous root policy digest is malformed"
      end
      {
        epoch: epoch,
        previous_policy_sha256: previous,
        roots: normalize_roots(field(policy, :roots)),
        signatures: normalize_signatures(field(policy, :signatures)),
        digest: digest(policy)
      }
    end

    def validate_chain(policies, pinned_bootstrap_roots, minimum_epoch: 0)
      unless policies.is_a?(Array) && !policies.empty?
        raise RootPolicyError, "root policy chain must be a nonempty Array"
      end
      raise RootPolicyError, "pinned roots must be a Hash" unless pinned_bootstrap_roots.is_a?(Hash)

      pinned = pinned_bootstrap_roots.to_h do |id, key|
        normalized_id = key_id(id)
        unless key.is_a?(String) && key.bytesize == 32
          raise RootPolicyError, "pinned bootstrap root must be 32 bytes"
        end
        [normalized_id, key.b]
      end
      previous = nil
      policies.each_with_index do |raw_policy, index|
        current = validate_shape(raw_policy)
        current_all = root_map(current[:roots])
        current_active = root_map(current[:roots], "active")
        if index.zero?
          unless current[:previous_policy_sha256].nil?
            raise RootPolicyError, "bootstrap policy must not have a predecessor"
          end
          unless current_active == pinned && current[:roots].length == pinned.length
            raise RootPolicyError, "bootstrap policy disagrees with pinned roots"
          end
          required_signers = pinned.keys
          verification_keys = pinned
        else
          unless current[:epoch] == previous[:epoch] + 1
            raise RootPolicyError, "root policy epoch must increase by exactly one"
          end
          unless current[:previous_policy_sha256] == previous[:digest]
            raise RootPolicyError, "root policy chain digest mismatch"
          end
          previous_all = root_map(previous[:roots])
          previous_active = root_map(previous[:roots], "active")
          previous_all.each do |id, old_key|
            unless current_all[id] == old_key
              raise RootPolicyError, "root key removal or replacement is forbidden"
            end
            old_status = previous[:roots].find { |entry| entry[:key_id] == id }[:status]
            new_status = current[:roots].find { |entry| entry[:key_id] == id }[:status]
            if old_status == "retired" && new_status != "retired"
              raise RootPolicyError, "retired root cannot be reactivated"
            end
          end
          required_signers = (previous_active.keys | current_active.keys)
          verification_keys = previous_all.merge(current_all)
        end
        signatures = current[:signatures].to_h do |entry|
          [entry[:key_id], entry[:signature_hex]]
        end
        unless signatures.keys.sort == required_signers.sort
          raise RootPolicyError, "root policy signatures do not match required signers"
        end
        required_signers.sort.each do |id|
          key = Authentication.ed25519_public_from_bytes(verification_keys.fetch(id))
          signed = [signatures.fetch(id)].pack("H*")
          unless !key.nil? && key.verify(nil, signed, message(raw_policy))
            raise RootPolicyError, "root policy signature is invalid"
          end
        rescue KeyError, OpenSSL::PKey::PKeyError
          raise RootPolicyError, "root policy signature is invalid"
        end
        previous = current
      end
      minimum = Generation.validate(minimum_epoch, "minimum root policy epoch")
      if previous[:epoch] < minimum
        raise RootPolicyError,
              "root policy epoch #{previous[:epoch]} is below minimum #{minimum}"
      end
      previous
    end

    def active_keys(validated_policy)
      root_map(validated_policy[:roots], "active")
    end

    def chain_checksum(chain)
      material = ROOT_STATE_DOMAIN.dup
      chain.each do |policy|
        material << [digest(policy)].pack("H*")
        normalize_signatures(field(policy, :signatures)).each do |signed|
          material << signed[:key_id].b
          material << [signed[:signature_hex]].pack("H*")
        end
      end
      Digest::SHA256.hexdigest(material)
    end
  end

  class RootPolicyStore
    attr_reader :slots, :minimum_epoch, :pinned_bootstrap_roots

    def initialize(bootstrap_policy, pinned_bootstrap_roots)
      bootstrap = Marshal.load(Marshal.dump(bootstrap_policy))
      validated = RootPolicy.validate_chain([bootstrap], pinned_bootstrap_roots)
      @pinned_bootstrap_roots = pinned_bootstrap_roots
      @slots = [
        RootStateSlot.new(chain: [bootstrap], checksum: RootPolicy.chain_checksum([bootstrap])),
        RootStateSlot.new
      ]
      @minimum_epoch = TrustedGenerationJournal.new(validated[:epoch])
    end

    def valid_slots
      @slots.each_with_index.filter_map do |slot, index|
        next if slot.chain.nil?

        begin
          next unless slot.checksum == RootPolicy.chain_checksum(slot.chain)

          validated = RootPolicy.validate_chain(
            slot.chain,
            @pinned_bootstrap_roots,
            minimum_epoch: @minimum_epoch.minimum_accepted
          )
        rescue RootPolicyError, KeyError, TypeError
          next
        end
        [index, validated]
      end
    end

    def current
      valid = valid_slots
      if valid.empty?
        raise RootPolicyError, "no valid root policy chain satisfies minimum epoch"
      end

      index, validated = valid.max_by { |_slot_index, candidate| candidate[:epoch] }
      [index, validated, @slots[index].chain]
    end

    def power_cut(fail_after)
      lambda do |step|
        raise SimulatedPowerLoss, step if step == fail_after
      end
    end

    def install(policy, fail_after: nil)
      after_step = power_cut(fail_after)
      current_index, _validated, chain = current
      candidate_chain = Marshal.load(Marshal.dump(chain))
      candidate_chain << Marshal.load(Marshal.dump(policy))
      RootPolicy.validate_chain(
        candidate_chain,
        @pinned_bootstrap_roots,
        minimum_epoch: @minimum_epoch.minimum_accepted
      )
      target = 1 - current_index
      @slots[target] = RootStateSlot.new
      after_step.call(ROOT_POLICY_INSTALL_STEPS[0])
      @slots[target].chain = candidate_chain
      after_step.call(ROOT_POLICY_INSTALL_STEPS[1])
      @slots[target].checksum = RootPolicy.chain_checksum(candidate_chain)
      after_step.call(ROOT_POLICY_INSTALL_STEPS[2])
      target
    end

    def commit_current(fail_after: nil)
      _index, validated, = current
      @minimum_epoch.commit(validated[:epoch], power_cut(fail_after))
    end
  end
end

File: workbench/min0_core_forth_root.py
"""Cross-signed root-policy chain and power-loss-safe state for MIN0 CORE FORTH."""

from __future__ import annotations

import copy
import hashlib
import json
import re
from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from min0_core_forth_generation import validate_generation
from min0_core_forth_install import SimulatedPowerLoss, TrustedGenerationJournal


ROOT_POLICY_FORMAT = "min0-core-forth-root-policy"
ROOT_POLICY_VERSION = 1
ROOT_POLICY_DOMAIN = b"MIN0-CORE-FORTH-ROOT-POLICY-R0\0"
ROOT_STATE_DOMAIN = b"MIN0-CORE-FORTH-ROOT-STATE-R0\0"
ROOT_POLICY_INSTALL_STEPS = (
    "erase-inactive-root-state",
    "write-root-policy-chain",
    "seal-root-state",
)
KEY_ID_PATTERN = re.compile(r"[a-z0-9][a-z0-9._-]{0,63}")


class RootPolicyError(ValueError):
    pass


def _key_id(value: object) -> str:
    if not isinstance(value, str) or KEY_ID_PATTERN.fullmatch(value) is None:
        raise RootPolicyError("root key_id is malformed")
    return value


def _public_key(value: object) -> bytes:
    if not isinstance(value, str) or len(value) != 64 or value.lower() != value:
        raise RootPolicyError("root public key must be 32 bytes of lowercase hex")
    try:
        return bytes.fromhex(value)
    except ValueError as exc:
        raise RootPolicyError("root public key hex is malformed") from exc


def _signature(value: object) -> bytes:
    if not isinstance(value, str) or len(value) != 128 or value.lower() != value:
        raise RootPolicyError("root policy signature must be 64 bytes of lowercase hex")
    try:
        return bytes.fromhex(value)
    except ValueError as exc:
        raise RootPolicyError("root policy signature hex is malformed") from exc


def _normalize_roots(roots: object) -> list[dict[str, str]]:
    if not isinstance(roots, Sequence) or isinstance(roots, (str, bytes, bytearray)):
        raise RootPolicyError("root entries must be a sequence")
    normalized = []
    seen = set()
    for entry in roots:
        if not isinstance(entry, Mapping) or set(entry) != {
            "key_id",
            "public_key_hex",
            "status",
        }:
            raise RootPolicyError("root entry fields are malformed")
        key_id = _key_id(entry.get("key_id"))
        if key_id in seen:
            raise RootPolicyError("duplicate root key_id")
        seen.add(key_id)
        public_key = _public_key(entry.get("public_key_hex"))
        status = entry.get("status")
        if status not in ("active", "retired"):
            raise RootPolicyError("root status must be active or retired")
        normalized.append(
            {
                "key_id": key_id,
                "public_key_hex": public_key.hex(),
                "status": status,
            }
        )
    normalized.sort(key=lambda item: item["key_id"])
    if not any(entry["status"] == "active" for entry in normalized):
        raise RootPolicyError("root policy must retain an active root")
    return normalized


def _normalize_signatures(signatures: object) -> list[dict[str, str]]:
    if not isinstance(signatures, Sequence) or isinstance(
        signatures, (str, bytes, bytearray)
    ):
        raise RootPolicyError("root signatures must be a sequence")
    normalized = []
    seen = set()
    for entry in signatures:
        if not isinstance(entry, Mapping) or set(entry) != {
            "key_id",
            "signature_hex",
        }:
            raise RootPolicyError("root signature entry fields are malformed")
        key_id = _key_id(entry.get("key_id"))
        if key_id in seen:
            raise RootPolicyError("duplicate root signature key_id")
        seen.add(key_id)
        normalized.append(
            {
                "key_id": key_id,
                "signature_hex": _signature(entry.get("signature_hex")).hex(),
            }
        )
    return sorted(normalized, key=lambda item: item["key_id"])


def _payload(policy: Mapping[str, object]) -> bytes:
    roots = _normalize_roots(policy.get("roots"))
    rows = [
        [entry["key_id"], entry["public_key_hex"], entry["status"]]
        for entry in roots
    ]
    payload = [
        policy.get("format"),
        policy.get("version"),
        policy.get("epoch"),
        policy.get("previous_policy_sha256"),
        rows,
    ]
    return json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode()


def policy_digest(policy: Mapping[str, object]) -> str:
    return hashlib.sha256(_payload(policy)).hexdigest()


def _message(policy: Mapping[str, object]) -> bytes:
    return ROOT_POLICY_DOMAIN + bytes.fromhex(policy_digest(policy))


def _root_map(roots: Sequence[Mapping[str, str]], status: str | None = None) -> dict[str, bytes]:
    return {
        entry["key_id"]: bytes.fromhex(entry["public_key_hex"])
        for entry in roots
        if status is None or entry["status"] == status
    }


def build_root_policy(
    epoch: int,
    roots: Sequence[Mapping[str, object]],
    signer_private_keys: Mapping[str, object],
    *,
    previous_policy: Mapping[str, object] | None = None,
) -> dict:
    normalized_epoch = validate_generation(epoch, "root policy epoch")
    normalized_roots = _normalize_roots(roots)
    policy = {
        "format": ROOT_POLICY_FORMAT,
        "version": ROOT_POLICY_VERSION,
        "epoch": normalized_epoch,
        "previous_policy_sha256": (
            policy_digest(previous_policy) if previous_policy is not None else None
        ),
        "roots": normalized_roots,
        "signatures": [],
    }
    signatures = []
    for key_id in sorted(signer_private_keys):
        normalized_key_id = _key_id(key_id)
        try:
            signed = signer_private_keys[key_id].sign(_message(policy))
        except (TypeError, ValueError, AttributeError) as exc:
            raise RootPolicyError("invalid root signer private key") from exc
        signatures.append(
            {"key_id": normalized_key_id, "signature_hex": signed.hex()}
        )
    policy["signatures"] = signatures
    return policy


def _validate_policy_shape(policy: object) -> dict:
    if not isinstance(policy, Mapping):
        raise RootPolicyError("root policy must be a mapping")
    if policy.get("format") != ROOT_POLICY_FORMAT:
        raise RootPolicyError("unsupported root policy format")
    if policy.get("version") != ROOT_POLICY_VERSION:
        raise RootPolicyError("unsupported root policy version")
    epoch = validate_generation(policy.get("epoch"), "root policy epoch")
    previous = policy.get("previous_policy_sha256")
    if previous is not None and (
        not isinstance(previous, str)
        or len(previous) != 64
        or previous.lower() != previous
    ):
        raise RootPolicyError("previous root policy digest is malformed")
    if previous is not None:
        try:
            bytes.fromhex(previous)
        except ValueError as exc:
            raise RootPolicyError("previous root policy digest is malformed") from exc
    roots = _normalize_roots(policy.get("roots"))
    signatures = _normalize_signatures(policy.get("signatures"))
    return {
        "epoch": epoch,
        "previous_policy_sha256": previous,
        "roots": roots,
        "signatures": signatures,
        "digest": policy_digest(policy),
    }


def validate_root_policy_chain(
    policies: object,
    pinned_bootstrap_roots: Mapping[str, object],
    *,
    minimum_epoch: int = 0,
) -> dict:
    if not isinstance(policies, Sequence) or isinstance(
        policies, (str, bytes, bytearray)
    ) or not policies:
        raise RootPolicyError("root policy chain must be a nonempty sequence")
    pinned = {}
    for key_id, public_key in pinned_bootstrap_roots.items():
        normalized_key_id = _key_id(key_id)
        if not isinstance(public_key, (bytes, bytearray, memoryview)) or len(public_key) != 32:
            raise RootPolicyError("pinned bootstrap root must be 32 bytes")
        pinned[normalized_key_id] = bytes(public_key)
    previous = None
    for index, raw_policy in enumerate(policies):
        current = _validate_policy_shape(raw_policy)
        current_all = _root_map(current["roots"])
        current_active = _root_map(current["roots"], "active")
        if index == 0:
            if current["previous_policy_sha256"] is not None:
                raise RootPolicyError("bootstrap policy must not have a predecessor")
            if current_active != pinned or len(current["roots"]) != len(pinned):
                raise RootPolicyError("bootstrap policy disagrees with pinned roots")
            required_signers = set(pinned)
            verification_keys = pinned
        else:
            assert previous is not None
            if current["epoch"] != previous["epoch"] + 1:
                raise RootPolicyError("root policy epoch must increase by exactly one")
            if current["previous_policy_sha256"] != previous["digest"]:
                raise RootPolicyError("root policy chain digest mismatch")
            previous_all = _root_map(previous["roots"])
            previous_active = _root_map(previous["roots"], "active")
            for key_id, previous_key in previous_all.items():
                if key_id not in current_all or current_all[key_id] != previous_key:
                    raise RootPolicyError("root key removal or replacement is forbidden")
                previous_status = next(
                    entry["status"]
                    for entry in previous["roots"]
                    if entry["key_id"] == key_id
                )
                current_status = next(
                    entry["status"]
                    for entry in current["roots"]
                    if entry["key_id"] == key_id
                )
                if previous_status == "retired" and current_status != "retired":
                    raise RootPolicyError("retired root cannot be reactivated")
            required_signers = set(previous_active) | set(current_active)
            verification_keys = {**previous_all, **current_all}
        signatures = {entry["key_id"]: entry["signature_hex"] for entry in current["signatures"]}
        if set(signatures) != required_signers:
            raise RootPolicyError("root policy signatures do not match required signers")
        for key_id in sorted(required_signers):
            try:
                Ed25519PublicKey.from_public_bytes(verification_keys[key_id]).verify(
                    bytes.fromhex(signatures[key_id]), _message(raw_policy)
                )
            except (InvalidSignature, ValueError, KeyError) as exc:
                raise RootPolicyError("root policy signature is invalid") from exc
        previous = current
    assert previous is not None
    minimum = validate_generation(minimum_epoch, "minimum root policy epoch")
    if previous["epoch"] < minimum:
        raise RootPolicyError(
            f"root policy epoch {previous['epoch']} is below minimum {minimum}"
        )
    return previous


def active_root_keys(validated_policy: Mapping[str, object]) -> dict[str, bytes]:
    return _root_map(validated_policy["roots"], "active")


def _chain_checksum(chain: Sequence[Mapping[str, object]]) -> str:
    material = bytearray(ROOT_STATE_DOMAIN)
    for policy in chain:
        material.extend(bytes.fromhex(policy_digest(policy)))
        for signature in _normalize_signatures(policy.get("signatures")):
            material.extend(signature["key_id"].encode("ascii"))
            material.extend(bytes.fromhex(signature["signature_hex"]))
    return hashlib.sha256(material).hexdigest()


@dataclass
class RootStateSlot:
    chain: list[dict] | None = None
    checksum: str | None = None


class RootPolicyStore:
    def __init__(
        self,
        bootstrap_policy: Mapping[str, object],
        pinned_bootstrap_roots: Mapping[str, object],
    ) -> None:
        bootstrap = copy.deepcopy(dict(bootstrap_policy))
        validated = validate_root_policy_chain([bootstrap], pinned_bootstrap_roots)
        self.pinned_bootstrap_roots = pinned_bootstrap_roots
        self.slots = [
            RootStateSlot([bootstrap], _chain_checksum([bootstrap])),
            RootStateSlot(),
        ]
        self.minimum_epoch = TrustedGenerationJournal(validated["epoch"])

    def _valid(self) -> list[tuple[int, dict]]:
        result = []
        for index, slot in enumerate(self.slots):
            if slot.chain is None:
                continue
            try:
                if slot.checksum != _chain_checksum(slot.chain):
                    continue
                validated = validate_root_policy_chain(
                    slot.chain,
                    self.pinned_bootstrap_roots,
                    minimum_epoch=self.minimum_epoch.minimum_accepted,
                )
            except (RootPolicyError, KeyError, TypeError):
                continue
            result.append((index, validated))
        return result

    def current(self) -> tuple[int, dict, list[dict]]:
        valid = self._valid()
        if not valid:
            raise RootPolicyError("no valid root policy chain satisfies minimum epoch")
        index, validated = max(valid, key=lambda item: item[1]["epoch"])
        return index, validated, self.slots[index].chain

    @staticmethod
    def _power_cut(fail_after: str | None) -> Callable[[str], None]:
        def after(step: str) -> None:
            if step == fail_after:
                raise SimulatedPowerLoss(step)

        return after

    def install(
        self, policy: Mapping[str, object], *, fail_after: str | None = None
    ) -> int:
        after = self._power_cut(fail_after)
        current_index, _validated, chain = self.current()
        candidate_chain = copy.deepcopy(chain) + [copy.deepcopy(dict(policy))]
        validate_root_policy_chain(
            candidate_chain,
            self.pinned_bootstrap_roots,
            minimum_epoch=self.minimum_epoch.minimum_accepted,
        )
        target = 1 - current_index
        self.slots[target] = RootStateSlot()
        after(ROOT_POLICY_INSTALL_STEPS[0])
        self.slots[target].chain = candidate_chain
        after(ROOT_POLICY_INSTALL_STEPS[1])
        self.slots[target].checksum = _chain_checksum(candidate_chain)
        after(ROOT_POLICY_INSTALL_STEPS[2])
        return target

    def commit_current(self, *, fail_after: str | None = None) -> int:
        _index, validated, _chain = self.current()
        return self.minimum_epoch.commit(
            validated["epoch"], self._power_cut(fail_after)
        )
